docs: document how to use the completion module (#326)
canonical/craft-cli: docs/how-to-guides/use-completion.rst

Reviewer: A minor-to-moderate point, but ideally the filename would match the natural name: `use-the-completion-module`

.. _use_completion:

Use the completion module
=========================

> Craft-cli provides a completion module for auto-generating bash completion scripts for applications using its :py:class:`Dispatcher`.

Reviewer 2: these lines are huge, I think our standard today is 88 characters per line

.. _use_completion:

Use the completion module
=========================

Craft-cli provides a completion module for auto-generating bash completion scripts for
applications using its :py:class:`Dispatcher`.

> In order to invoke it, the application needs to have a public function that returns

Author: The module itself is called "completion" - what do you recommend?

> In order to invoke it, the application needs to have a public function that returns

Reviewer: Ah, that makes more sense. Workaround: https://github.com/canonical/craft-cli/pull/326#discussion_r1955338602

some basic information about itself. All applications must provide their
:py:class:`Dispatcher` and a configuration to initialise commands with. By default,
craft-cli commands don't need to be initialised with anything, so this would be

> ``None`` in the basic case.

Reviewer: It took me a while to realise the configuration we're referring to is the `None` in the function's return.

I think we could make this segment clearer by:
- Clarifying what we're going to do with this app getter.
- Discussing more plainly what the developer will do with the configuration.
- In the absence of a reference, describe the relationship between the module and the dispatcher.

> ``None`` in the basic case.

Author: I think this looks good - thanks!

> For a project named "testcraft", create the file :file:`testcraft/application.py` and

Reviewer: ```suggestion
For a project named Testcraft, create the file :file:`testcraft/application.py` and
```

I try and avoid quotation marks.

add the following content:

.. code:: python

    from craft_cli import Dispatcher

    def get_dispatcher() -> Dispatcher:
        """Fill out this function to create the application's dispatcher"""

    def get_app_info() -> tuple[Dispatcher, None]:
        return get_dispatcher(), None

> Once the function is made, the completion module can be invoked to create the

Reviewer: I think this should be its own section.

completion file:

.. code:: shell

    python3 -m craft_cli.completion testcraft testcraft.application:get_app_info > completion.sh

Applications using craft-application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Commands from `craft-application`_ need their ``app_config`` dictionary in order to be

Reviewer: Maybe an example of an app config? That could go very far in this doc.

Reviewer 3: The `app_config` dictionary is a property generated by the `Application` object itself, and the correct way to access it is what's shown on line 56 (accessing the `app_config` property of the `Application` object).

I think a better way to clarify this would be to provide an example implementation of `get_app()` that returns the `Application`. I'll write an example below.

Reviewer 3: https://github.com/canonical/craft-cli/pull/326#discussion_r1953384802

Reviewer: Thanks! Some of the details escape me but I find your code has better signposting for users.

initialised. To handle this, return the ``app_config`` alongside the application's

> :py:class:`Dispatcher` object:

Reviewer 3: ```suggestion
:py:class:`~craft_cli.dispatcher:Dispatcher` object:
```

.. code:: python

>     from craft_cli import Dispatcher
>     from typing import Any
> 
>     def get_app() -> MyApplication:
>         """Fill out this function to create an application"""
> 
>     def get_dispatcher() -> Dispatcher:
>         """Fill out this function to create the application's dispatcher"""

Reviewer 3: ```suggestion
    from craft_application import Application, ServiceFactory
    from craft_cli import Dispatcher
    from typing import Any
    
    from my_app import APP_METADATA, commands

    def get_app() -> Application:
        """Replace this function with a function that instantiates your app."""
        services = ServiceFactory(app=APP_METADATA)
        app = Application(app=APP_METADATA, services=services)
        commands.fill_command_groups(app)
        return app

    def get_dispatcher() -> Dispatcher:
        """Fill out this function to create the application's dispatcher"""
```

.. _use_completion:

Use the completion module
=========================

> Craft CLI provides a completion module for auto-generating Bash completion scripts for

Reviewer: ```suggestion
Craft CLI provides the completion module, which auto-generates Bash completion scripts for
```

apps.

Write the app info getter
-------------------------

To invoke the module, an app needs to have a public function that returns some basic

> information about itself. The app must provide a Dispatcher with a configuration to

Reviewer: ```suggestion
information about itself. The app must provide a :py:class:`Dispatcher` with a configuration to
```

The Dispatcher is where the commands themselves are pulled in and transformed into
entries for the Bash script. The commands inside the Dispatcher are initialised and
then parsed for their options and inputs.

The purpose of the getter is to give the module an entry point into your application
for it to gather the necessary information to build a completion script.

For a project named Testcraft, create the file :file:`testcraft/application.py` and
add the following content:

.. code:: python

    from craft_cli import Dispatcher

    def get_dispatcher() -> Dispatcher:
        """Fill out this function to create the application's dispatcher"""

    def get_app_info() -> tuple[Dispatcher, None]:
        # Returning the Dispatcher and no initialisation config
        return get_dispatcher(), None

Create the completion script
----------------------------

Once the function is made, the completion module can be invoked to create the
completion file:

.. code:: shell

    python3 -m craft_cli.completion testcraft testcraft.application:get_app_info > completion.sh

Applications using craft-application
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Commands from `craft-application`_ need their corresponding ``Application``'s

Reviewer: This might reduce the inline code overhead:

```suggestion
Commands from `craft-application`_ need their corresponding ``Application`` object's
```
